tests/flytekit/unit/extras/tasks/test_shell.py

```diff
@@ -370,3 +370,27 @@ def test_subproc_execute_with_shell():
         subproc_execute(cmd, shell=True)
         cont = open(opth).read()
         assert "hello" in cont
+
+
+def test_subproc_execute_missing_dep():
+    cmd = ["non-existent", "blah"]
+    with pytest.raises(Exception) as e:
+        subproc_execute(cmd)
+    assert "executable could not be found" in str(e.value)
+
+
+def test_subproc_execute_error():
+    cmd = ["ls", "--banana"]
+    with pytest.raises(Exception) as e:
+        subproc_execute(cmd)
+    assert "Failed with return code" in str(e.value)
+
+
+def test_subproc_execute_shell_error(tmp_path):
+    # This is a corner case I ran into that really shouldn't
+    # ever happen. The assert catches anything in stderr despite
+    # a 0 exit.
+    cmd = " ".join(["bcftools", "isec", "|", "gzip", "-c", ">", f"{tmp_path.joinpath('out')}"])
```

Comment:
the exitcode `0` only happens because the second command (i.e. `gzip`) returns a `0`. Bash implements a feature called PIPESTATUS, which captures the exit codes of all commands in a pipe. It's just an array that you can verify after the pipe is run, for example:
```
bash-5.2$ inexistet command --flag | echo bla | gzip -c > out
bash: inexistet: command not found
bash-5.2$ echo "${PIPESTATUS[@]}"
127 0 0
```
Notice how the first command returns `127` and the other two return `0`.

Reply:
TIL, thank you!